### DISCUSSION TITLE
Is it fine to have long living inserts?

### DISCUSSION THREAD

**Discussion by nikandfor** [ClickHouse/ClickHouse#50263] [Q&A] [answered]
This is example Go code. Is it fine for the server, or it will cause delays and other problems for other clients?
```
b, err := conn.PrepareBatch(ctx, `INSERT INTO table (colA, colB) VALUES`)
// if err != nil ...

defer func() {
    e := b.Send()
    // if err == nil && e != nil ...
}()

t := time.NewTimer(10 * time.Second)
defer t.Stop()

var row Row

for {
    select {
    case row = <-rowsChan:
    case <-t.C:
        return nil
    case <-ctx.Done():
        return nil
    }
    
    err = b.Append(row.ColA, row.ColB)
    // if err != nil ...
}
```

And to not bother you twice, is it fine to do the same for reader? So If I have big query result and I consume it slowly, will it produce any problems for the server and the rest of clients? I mean connections and memory usage is obvious, but is there anything else?

**Comment by tbragin** [comment, marked as answer]
One other downside I can think of for long-running insert or select jobs is network interruptions, but as long as your server has sufficient resources to process a long-running insert or select you should be fine. 

For context, my main experience with this is what we are seeing in ClickHouse cloud - we see some long-running inserts (usually large data onboarding jobs) in ClickHouse Cloud take hours and even days in some cases. But we also usually advise customers at that point to break them up into smaller jobs to protect against network problems.

**Reply by nikandfor** [reply]
Thank you for the reply. One more little question. In case of long insert what happen to the data already sent to the server if connection breaks? As I can guess it will be discarded, am I right?

**Reply by tbragin** [reply]
The query will get interrupted and the insert will be partially complete. It is hard to identify the exact point of failure, but thankfully there are tricks to take care of that. @gingerwizard just told me that he usually uses mod of an int column or hash of a string column, and then uses a  blocking lightweight delete to remove the failed subset before continuing. I asked him to see if he can put an example together and share here.

**Reply by gingerwizard** [reply]
Hi @nikandfor typically i use a simple modulus of a int column with decent cardinality i.e. 
```sql
INSERT INTO X SELECT * FROM table WHERE column % 100 = 0
INSERT INTO X SELECT * FROM table WHERE column % 100 = 1
INSERT INTO X SELECT * FROM table WHERE column % 100 = 2
...

INSERT INTO X SELECT * FROM table WHERE column % 100 = 99
```

In the absence of an integer column, we can  also use a [CityHash](https://clickhouse.com/docs/en/sql-reference/functions/hash-functions#cityhash64) e.g.

```sql
INSERT INTO X SELECT * FROM table WHERE cityHash64(column) % 100 = 0
etc.
```

In the event of failure you can delete the batch which failed using a _[lightweight delete](https://clickhouse.com/docs/en/guides/developer/lightweght-delete)_, blocking ([either monitor](https://clickhouse.com/docs/en/operations/system-tables/mutations#monitoring-mutations) or [block](https://clickhouse.com/docs/en/operations/settings/settings#mutations_sync)) until it completes. Other batches can proceed but don't restart the failed batch until this delete has been completed.
